File: frontend/routers/items.py
from datetime import datetime
from fastapi import APIRouter, HTTPException, UploadFile, File
from typing import List, Dict, Any
from ..models.schemas import Item, ItemCreate, ItemUpdate
from ..services.firestore import FirestoreService
from ..services.ml import PredictionService

router = APIRouter()


def get_db() -> FirestoreService:
    # Lazy initialization to avoid import-time failures when Firestore is not available
    return FirestoreService()


# Simple in-memory fallback storage for local dev when Firestore/emulator is unavailable
_INMEM_DB: Dict[str, Dict[str, Any]] = {}


def get_ml() -> PredictionService:
    return PredictionService()


@router.get("/users/{user_id}/items", response_model=List[Item])
async def list_items(user_id: str) -> List[Item]:
    try:
        db = get_db()
        return await db.list_items(user_id)
    except Exception:
        # Graceful fallback in local dev if Firestore/emulator isn't running
        user_items = _INMEM_DB.get(user_id, {})
        items: List[Item] = []
        for iid, data in user_items.items():
            data_with_id = dict(data)
            data_with_id["id"] = iid
            items.append(Item(**data_with_id))
        return items


@router.post("/classify-image")
async def classify_image(file: UploadFile = File(...)) -> dict:
    """Lightweight demo endpoint: suggests name/category from the image filename
    and returns an expiry/status using the existing PredictionService.

    This is a stub for local development. Replace with a real vision model later.
    """
    try:
        filename = (file.filename or "").lower()
        # Very naive heuristics by filename keywords
        guess_map = {
            "apple": ("Apple", "fruits"),
            "banana": ("Banana", "fruits"),
            "milk": ("Milk", "dairy"),
            "bread": ("Bread", "bakery"),
            "tomato": ("Tomato", "vegetable"),
            "tomatoe": ("Tomato", "vegetable"),
            "carrot": ("Carrot", "vegetable"),
            "chicken": ("Chicken", "meat"),
            "fish": ("Fish", "meat"),
            "yogurt": ("Yogurt", "dairy"),
            "cheese": ("Cheese", "dairy"),
        }
        name = "Food Item"
        category = None
        for key, (n, c) in guess_map.items():
            if key in filename:
                name, category = n, c
                break

        # Run our existing prediction logic (storage default fridge)
        pred = await get_ml().predict_expiry({
            "name": name,
            "category": category,
            "storage_condition": "fridge",
        })

        return {
            "name": name,
            "category": category,
            "storage_condition": "fridge",
            "predicted_expiry": pred.predicted_expiry,
            "status": pred.status,
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to classify image: {e}")


@router.post("/users/{user_id}/items", response_model=Item)
async def create_item(user_id: str, payload: ItemCreate) -> Item:
    try:
        ml = get_ml()
        db = get_db()
        item_dict = payload.dict()
        prediction = await ml.predict_expiry(item_dict)
        item_dict["predicted_expiry"] = prediction.predicted_expiry
        item_dict["status"] = prediction.status
        item_dict["created_at"] = datetime.utcnow()
        item_dict["updated_at"] = datetime.utcnow()
        created = await db.create_item(user_id, item_dict)
        return created
    except Exception:
        # In-memory fallback
        from uuid import uuid4
        item_id = str(uuid4())
        item_dict = payload.dict()
        prediction = await get_ml().predict_expiry(item_dict)
        item = Item(
            id=item_id,
            name=item_dict["name"],
            category=item_dict.get("category"),
            storage_condition=item_dict.get("storage_condition"),
            opened=item_dict.get("opened", False),
            predicted_expiry=prediction.predicted_expiry,
            status=prediction.status,
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow(),
            purchase_date=item_dict.get("purchase_date"),
        )
        _INMEM_DB.setdefault(user_id, {})[item_id] = item.dict()
        return item


@router.get("/users/{user_id}/items/{item_id}", response_model=Item)
async def get_item(user_id: str, item_id: str) -> Item:
    try:
        db = get_db()
        item = await db.get_item(user_id, item_id)
        if not item:
            raise HTTPException(status_code=404, detail="Item not found")
        return item
    except HTTPException:
        raise
    except Exception:
        user_items = _INMEM_DB.get(user_id, {})
        data = user_items.get(item_id)
        if not data:
            raise HTTPException(status_code=404, detail="Item not found")
        return Item(**data)


@router.patch("/users/{user_id}/items/{item_id}", response_model=Item)
async def update_item(user_id: str, item_id: str, payload: ItemUpdate) -> Item:
    try:
        db = get_db()
        update_data = {k: v for k, v in payload.dict(exclude_unset=True).items()}
        update_data["updated_at"] = datetime.utcnow()
        updated = await db.update_item(user_id, item_id, update_data)
        if not updated:
            raise HTTPException(status_code=404, detail="Item not found")
        return updated
    except HTTPException:
        raise
    except Exception:
        # In-memory fallback
        user_items = _INMEM_DB.setdefault(user_id, {})
        existing = user_items.get(item_id)
        if not existing:
            raise HTTPException(status_code=404, detail="Item not found")
        existing.update({k: v for k, v in payload.dict(exclude_unset=True).items()})
        existing["updated_at"] = datetime.utcnow()
        return Item(**existing)


@router.delete("/users/{user_id}/items/{item_id}")
async def delete_item(user_id: str, item_id: str) -> dict:
    try:
        db = get_db()
        deleted = await db.delete_item(user_id, item_id)
        if not deleted:
            raise HTTPException(status_code=404, detail="Item not found")
        return {"deleted": True}
    except HTTPException:
        raise
    except Exception:
        # In-memory fallback
        user_items = _INMEM_DB.setdefault(user_id, {})
        if item_id not in user_items:
            raise HTTPException(status_code=404, detail="Item not found")
        del user_items[item_id]
        return {"deleted": True}

File: backend-express/routes/items.js
import express from 'express';
import { v4 as uuidv4 } from 'uuid';
import { verifyToken } from './auth.js';

const router = express.Router();

// In-memory items storage (replace with database in production)
const items = new Map();

// Helper function to predict expiry
function predictExpiry(item) {
  const expiryDays = {
    'fridge': { 'dairy': 7, 'vegetable': 5, 'fruit': 5, 'meat': 3, 'default': 5 },
    'freezer': { 'default': 90 },
    'pantry': { 'default': 30 }
  };

  const storage = item.storage_condition || 'fridge';
  const category = item.category?.toLowerCase() || 'default';
  const days = expiryDays[storage]?.[category] || expiryDays[storage]?.['default'] || 5;

  const purchaseDate = item.purchase_date ? new Date(item.purchase_date) : new Date();
  const expiryDate = new Date(purchaseDate);
  expiryDate.setDate(expiryDate.getDate() + days);

  const today = new Date();
  const daysUntilExpiry = Math.ceil((expiryDate - today) / (1000 * 60 * 60 * 24));

  let status = 'fresh';
  if (daysUntilExpiry < 0) {
    status = 'expired';
  } else if (daysUntilExpiry <= 3) {
    status = 'expiring soon';
  }

  return {
    predicted_expiry: expiryDate.toISOString().split('T')[0],
    status,
    daysUntilExpiry
  };
}

// Get all items for a user
router.get('/:userId/items', (req, res) => {
  try {
    const { userId } = req.params;
    const userItems = Array.from(items.values()).filter(item => item.userId === userId);
    res.json(userItems);
  } catch (error) {
    res.status(500).json({ error: 'Failed to fetch items', message: error.message });
  }
});

// Create item
router.post('/:userId/items', (req, res) => {
  try {
    const { userId } = req.params;
    const { name, category, storage_condition, opened, purchase_date } = req.body;

    if (!name) {
      return res.status(400).json({ error: 'Item name is required' });
    }

    const itemId = uuidv4();
    const newItem = {
      id: itemId,
      userId,
      name,
      category: category || null,
      storage_condition: storage_condition || 'fridge',
      opened: opened || false,
      purchase_date: purchase_date || new Date().toISOString().split('T')[0],
      createdAt: new Date().toISOString()
    };

    // Add expiry prediction
    const prediction = predictExpiry(newItem);
    newItem.predicted_expiry = prediction.predicted_expiry;
    newItem.status = prediction.status;

    items.set(itemId, newItem);

    res.status(201).json(newItem);
  } catch (error) {
    res.status(500).json({ error: 'Failed to create item', message: error.message });
  }
});

// Update item status
router.patch('/:userId/items/:itemId', (req, res) => {
  try {
    const { userId, itemId } = req.params;
    const { status } = req.body;
    
    const item = items.get(itemId);
    if (!item) {
      return res.status(404).json({ error: 'Item not found' });
    }

    if (item.userId !== userId) {
      return res.status(403).json({ error: 'Unauthorized' });
    }

    if (status) {
      item.status = status;
    }

    items.set(itemId, item);
    res.json(item);
  } catch (error) {
    res.status(500).json({ error: 'Failed to update item', message: error.message });
  }
});

// Delete item
router.delete('/:userId/items/:itemId', (req, res) => {
  try {
    const { userId, itemId } = req.params;
    
    const item = items.get(itemId);
    if (!item) {
      return res.status(404).json({ error: 'Item not found' });
    }

    if (item.userId !== userId) {
      return res.status(403).json({ error: 'Unauthorized' });
    }

    items.delete(itemId);
    res.json({ message: 'Item deleted successfully' });
  } catch (error) {
    res.status(500).json({ error: 'Failed to delete item', message: error.message });
  }
});

export default router;
